rfcs/20190610-standartizing-composite_ops.md

```diff
@@ -0,0 +1,323 @@
+provide efficient implementations for their target platform and break everyone
+else? Or should they provide tensorflow based graph implementation and then rely
+on graph pattern recognition to extract and match in the tooling of their
+choice.
+
+Who is affected? Tensorflow users, software and hardware vendors.
+
+### Existing prior art.
+
+Tensorflow Lite have developed [tf.lite.OpHint](https://www.tensorflow.org/api_docs/python/tf/lite/OpHint),
+which solves a very similar problem. However it is positioned as tf.lite
+specific extension, which doesn't provide a public api for graph consumers
+(other than TOCO) to extract the hints, limiting the potential for broader
+adoption by other tooling and limiting its usefulness to the users.
+
+`tf.lite.OpHint` also adds wrong direction to the flow
+of dependencies from tensorflow to lite, should core tensorflow api choose to 
+use to annotate composite ops.
+
+Pattern matching is another approach that is currently used. For example TfLite
+has pattern matching code that tries to detect various patterns that can then be
+converted into specialized ops, such as
+[prelu](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/toco/graph_transformations/identify_prelu.cc)
+and
+[lstm](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/toco/graph_transformations/identify_lstm.cc)
+and others files in tf.lite directory.
+
+## User Benefit
+
+The main beneficiaries would be:
+
+a) ML community that get a clean way of defining of composite ops without having
+to commit to particular implementation (or whether to build new `tf_op`)
+
+b) Tool maintainers that wouldn't need to write complicated and brittle graph
+extraction code whose sole purpose is to reverse engineer tensorflow
+implementations. For example here are some tf-lite transformations that identifies
+composite ops like [prelu], [lstm], [l2 pooling], etc. Which essentially
+requires updates to lite whenever tensorflow modifies those implementations.
+
+_Blog post announcement_: "How to add new functions to tensorflow"
+
+_Subtitle_: "So you don't regret it later".
+
+## Design Proposal
+
+Tensorflow already provides a mechanism to define functions inside a graph
+called `tf.function`. In short, `tf.function` allows to define a subgraph with
+dedicated inputs and outputs, that then is substituted as a custom node into
+graphs. The gist of our proposal is to add a few standardized attributes to
+signal to the processing/executing library that this function implements a
+“standard” function. The consumer library then can choose to process it
+differently should they choose to. The important part is that we define a
+standard set of attributes that users and tooling can rely on.
+
+As a start, we propose to add “implements” and “reference” as two attributes
+defining a “standard” function.
+
+For example:
+
+```
+@function.library_function(
+    implements=“google.matmul_low_rank_matrix”,
```

Comment:
For any composite ops moved from contrib would you track down the original developers or will everything in core fall under google namespace?